Add test addon to deploy rook based CephFS instance

ShyamsundarR opened the pull request:
... and required storage and snapshot classes to consume the same.

ShyamsundarR commented:
Effort was to test volsync with drenv, and hence adding a CephFS instance to drenv.

Ideally this should be invoked form the regional-dr env file, which I used to test with locally. Just waiting to resolve volsync [mover permission](https://volsync.readthedocs.io/en/latest/usage/permissionmodel.html) correctly before integrating this with the regional env.

An example ocm-sample that was tested is [here](https://github.com/ShyamsundarR/ocm-ramen-samples/commit/fc6801a56ce6cb8bb333e6ce3337bc6bc01341d6) for reference. This required that the `busybox-sample` namespace be annotated with `kubectl annotate ns/busybox-sample volsync.backube/privileged-movers=true` on both clusters for the movers to work.

nirs commented on test/addons/rook-cephfs/classes.yaml:
```diff
@@ -0,0 +1,41 @@
+allowVolumeExpansion: true
+---
+# SPDX-FileCopyrightText: The RamenDR authors
+# SPDX-License-Identifier: Apache-2.0
```
We don't need to repeat the SPDX headers, the are for the entire file, not for
the yaml documents.

But it would be nicer to have one file per resource, makes it easier to understand.

ShyamsundarR replied:
ack!

ShyamsundarR replied:
done.

nirs commented on test/addons/rook-cephfs/filesystem.yaml:
```diff
@@ -0,0 +1,27 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# yamllint disable-line rule:line-length
+# Source: https://raw.githubusercontent.com/rook/rook/release-1.13/deploy/examples/filesystem-test.yaml
```
Any reason not to use the remote yaml and patch it like we do with other rook resources?

This make the changes explicit - with this we have to download and diff the yaml to find whats was changed, and there is no comment about changed settings.

ShyamsundarR replied:
The YAML in rook also has a subvolumegroup resource, which I was trying to avoid. (single file in rook)

ShyamsundarR replied:
Added the comment on modification...

nirs reviewed (commented):
Nice! it would be even better with self test.

nirs commented:
> An example ocm-sample that was tested is [here](https://github.com/ShyamsundarR/ocm-ramen-samples/commit/fc6801a56ce6cb8bb333e6ce3337bc6bc01341d6) for reference. This required that the `busybox-sample` namespace be annotated with `kubectl annotate ns/busybox-sample volsync.backube/privileged-movers=true` on both clusters for the movers to work.

Annotating the busybox-sample namespace did not work when I tried, it is the namespace
created on the hub. OCM does not copy the annotations from the hub namespace to the manged clusters.

I tried to add a namespace to the application, but in this case OCM removes it during
cleanup.

I think we can solve this by adding the annotations in ramen on the manifestwork for the namespace, maybe only on k8s?

For real clusters I don't think we should promote running as root, instead we should
configure replicationsource/destination with the right security context, copied from
the pod.

ShyamsundarR commented:
> Nice! it would be even better with self test.

I did not add a self test because initially/originally the tests were intended for some components that were unstable. If we are adding tests to all addons, we  could add a PVC create/delete test against CephFS as a test.

ShyamsundarR commented:
> > Nice! it would be even better with self test.
> 
> I did not add a self test because initially/originally the tests were intended for some components that were unstable. If we are adding tests to all addons, we could add a PVC create/delete test against CephFS as a test.

Test added

ShyamsundarR commented on test/addons/rook-cephfs/test:
```diff
@@ -0,0 +1,72 @@
+        context=cluster,
+    )
+
+    print(f"CephFS based PVC and Snapshot provisioning on cluster '{cluster}' succeeded")
```
So there is an 88 char line length limitation in the python files? I see [this](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#labels-line-length) from black, which suggests so.

@nirs Considering we have a longer line length limitation overall in the project for the go files, I suggest we change this and add the `--line-lenght` parameter to the run of black? Having different line length requirement across does not make it intuitive when writing code.

nirs replied:
I don't think it is a problem, people using python expects the black default line length.

When using black you don't need to think about formatting or line length. You configure your editor to run Black on save (see .vimrc.example), and Black format the file for you. There is no need to discuss formatting issues.

Another solution is editorconfig (https://editorconfig.org/). We can add editor config file to the project specifying how various files and directories should be formatted, and the editors use the config to format.

If we want common line length for the entire project, we cannot use the current Go default (120) because it is bad for readability. I think Black default of 88 is good compromise so I'm happy to adopt it as the project line length.

ShyamsundarR replied:
I believe we should have a line length for the project not parts of it.

We are on 120 and I think we continue that, as before I am not debating line length, it is 120 for this project!

For now I am backing off changing the black defaults, but if this becomes an issue repeatedly I will change the default for black.

raghavendra-talur commented on test/addons/rook-cephfs/storage-class.yaml:
```diff
@@ -0,0 +1,24 @@
+  csi.storage.k8s.io/node-stage-secret-name: rook-csi-cephfs-node
+  csi.storage.k8s.io/node-stage-secret-namespace: rook-ceph
+reclaimPolicy: Delete
+allowVolumeExpansion: true
```
Do we want to add volume binding to be immediate or is that the default?

ShyamsundarR replied:
Upstream Ceph-csi examples also do not carry it, but unsure of binding default, maybe @Madhu-1 may know.

nirs commented on test/addons/rook-cephfs/test:
```diff
@@ -0,0 +1,72 @@
+    print(f"Deploying pvc {PVC_NAME} in cluster '{cluster}'")
+    kubectl.apply(
+        f"--filename={PVC_NAME}.yaml",
+        "--namespace=rook-ceph",
```
Maybe use a namespace for the test, so if the test fails, we don't leave leftovers in the rook namespace? 

I think we have the same issue in rbd-mirror, but lets not add more of these issues.

ShyamsundarR replied:
Done

nirs commented on test/addons/rook-cephfs/test:
```diff
@@ -0,0 +1,72 @@
+        "--namespace=rook-ceph",
+        "--timeout=300s",
+        context=cluster,
+    )
```
Would be useful to use the bounded PVC - for example start a busybox pod with it, so we know that we can read and write to it.

ShyamsundarR replied:
Why? Firstly these tests are not for every component, we do not do the same with RBD volumes, it is a sanity for the component that is added. I am unsure how much testing we want to do here and for what reason.

nirs replied:
The fact we don't test that rbd pvc works is also bad, but so far it worked for us. We can use the volsync test for validation (after switching to cephfs instead of csi-hostpath-driver).

nirs commented on test/addons/rook-cephfs/test:
```diff
@@ -0,0 +1,72 @@
+        "--namespace=rook-ceph",
+        "--timeout=300s",
+        context=cluster,
+    )
```
Cool to test snapshot, but how do we know that it is actually working? Could be useful to make change so the PVC, verify that they are available in the snapshot, and that new changes in the PVC are not.

ShyamsundarR replied:
We are not testing CephFS snapshots or PVC we are testing if what we installed is functional. This request seems to be more than what is needed to setup the environment.

nirs replied:
This is what we do for the volsync test, it will help to debug issues when we know that the cephfs test works if ramen does not, and fail earlier before we try to install ramen if cephfs pvc or snapshot do not work.

nirs commented on test/envs/regional-dr.yaml:
```diff
@@ -32,6 +32,7 @@ templates:
           - name: rook-cluster
           - name: rook-toolbox
           - name: rook-pool
+          - name: rook-cephfs
```
How much time this takes?

nirs replied:
OK, found in the e2e job:

```
2024-05-09 01:47:02,857 INFO    [rdr-dr1/0] addons/rook-cephfs/start completed in 39.78 seconds
2024-05-09 01:47:06,327 INFO    [rdr-dr1/0] addons/rook-cephfs/test completed in 3.47 seconds
```

Looks good. I wonder if we should remove the csi-hostpath-driver and volumesnapshot addons which slow down the run by minutes (part of time to start the cluster):

2024-05-09 01:39:21,265 INFO    [rdr-hub] Cluster started in 70.08 seconds
2024-05-09 01:40:25,387 INFO    [rdr-dr2] Cluster started in 134.20 seconds
2024-05-09 01:41:21,252 INFO    [rdr-dr1] Cluster started in 190.06 seconds

Without these addons the clusters all cluster can start in ~90 seconds.

We can work on this later.

ShyamsundarR replied:
On memory usage only inspection and time will tell, I did not face any issues in my usual runs (about 10-12 till now)

ShyamsundarR commented on test/addons/rook-cephfs/start:
```diff
@@ -0,0 +1,49 @@
+        "--filename=filesystem.yaml",
+        "--filename=storage-class.yaml",
+        "--filename=snap-class.yaml",
+        context=cluster,
```
Is it mandatory? Again modeled as per rook-pool addon, I do not see the need for a kustomize in this case.

nirs replied:
Ok, it looks very similar to rook-pool.

ShyamsundarR commented on test/addons/rook-cephfs/snap-class.yaml:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# yamllint disable-line rule:line-length
+# Source: https://raw.githubusercontent.com/rook/rook/release-1.13/deploy/examples/csi/cephfs/snapshotclass.yaml
```
Same reason as before, modeled around the rook-pool addon.

Same answer as before, source is provided for a diff in the future.

Are you insisting that what the delta is SHOULD be a comment in the YAML, or are you stating it should be used directly from Rook if not different? (I would need a common Rook version across the add-ons to do the latter)

(applies to all comments of this nature on sourced YAMLs)

nirs replied:
I'm thinking how we are going to update this when new rook version is released. We will need to diff the new rook config, and then understand the changes from our config vs the old rook config, and guess what was the indention of the author. Then think if we want to change something based on the changes in rook.

It will be easier it the yaml will document the intentions:
- Why we removed stuff
- Which values were changed, and why

In rook operator and cluster this is easy, we use kustomization and every change we made is documented.

If we don't change anything, using the origianl from rook is best, since we support caching kustomzizations.

We don't have global rook version for now, we duplicate the version in the kustomization yamls. I did not find a solution for keeping a version external to the kustomization, maybe you have an idea how to do this?

ShyamsundarR commented on test/addons/rook-cephfs/start:
```diff
@@ -0,0 +1,49 @@
+        "--namespace=rook-ceph",
+        "--timeout=300s",
+        context=cluster,
+    )
```
CephFilesystem status has an [observedGeneration](https://github.com/rook/rook/blob/5e3d62308c16274f32ca9cbfd73800292a95ed7c/pkg/apis/ceph.rook.io/v1/types.go#L1265)

The output has no conditions, e.g:
```
  status:
    observedGeneration: 1
    phase: Ready
```

So the current scheme should suffice?

nirs replied:
OK, so we don't have any other way.

ShyamsundarR commented:
@nirs I am done addressing comments that I think need addressing in this PR. I added a final commit to source from rook git instead of local copy of the resources.

ShyamsundarR commented on test/addons/rook-cephfs/provision-test/namespace.yaml:
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: busybox
```
Ugh! did not commit this file... (updating)

ShyamsundarR replied:
(updated)
